--- libunwind/src/ptrace/_UPT_access_fpreg.c	2015-08-18 10:59:40.000000000 +0300
+++ patches/_UPT_access_fpreg.c	2015-08-18 10:57:37.000000000 +0300
@@ -46,8 +46,8 @@
 #ifdef HAVE_TTRACE
 #       warning No support for ttrace() yet.
 #else
-        ptrace (PTRACE_POKEUSER, pid, _UPT_reg_offset[reg] + i * sizeof(wp[i]),
-                wp[i]);
+        ptrace (PTRACE_POKEUSER, pid, (void*) (_UPT_reg_offset[reg] + i * sizeof(wp[i])),
+                (void*) wp[i]);
 #endif
         if (errno)
           return -UNW_EBADREG;
@@ -59,7 +59,7 @@
 #       warning No support for ttrace() yet.
 #else
         wp[i] = ptrace (PTRACE_PEEKUSER, pid,
-                        _UPT_reg_offset[reg] + i * sizeof(wp[i]), 0);
+                        (void*) (_UPT_reg_offset[reg] + i * sizeof(wp[i])), 0);
 #endif
         if (errno)
           return -UNW_EBADREG;
@@ -100,6 +100,14 @@
 #endif
   return 0;
 }
+#elif defined(__aarch64__)
+int
+_UPT_access_fpreg (unw_addr_space_t as, unw_regnum_t reg, unw_fpreg_t *val,
+           int write, void *arg)
+{
+# pragma message("_UPT_access_fpreg is not implemented and not currently used.")
+  return -UNW_EBADREG;
+}
 #else
 #error Fix me
 #endif
